django/utils/datastructures.py

```diff
@@ -280,3 +281,86 @@ def __getitem__(self, key):
         if use_func:
             return self.func(value)
         return value
+
+
+def _destruct_iterable_mapping_values(data):
+    for i, elem in enumerate(data):
+        if len(elem) != 2:
+            raise ValueError("dictionary update sequence element #{} has "
+                             "length {}; 2 is required".format(i, len(elem)))
+
+        if not isinstance(elem[0], str):
+            raise ValueError('Element key invalid, only strings are allowed')
+
+        yield tuple(elem)
+
+
+class ImmutableCaseInsensitiveDict(Mapping):
+    """An immutable case-insensitive dictionary that still preserves
+    the case of the original keys used to create it."""
+
+    def __init__(self, data):
+        if not isinstance(data, Mapping):
+            data = {k: v for k, v in _destruct_iterable_mapping_values(data)}
+        self._store = {k.lower(): (k, v) for k, v in data.items()}
+
+    def __getitem__(self, key):
+        return self._store[key.lower()][1]
+
+    def __len__(self):
+        return len(self._store)
+
+    def __eq__(self, other):
+        if not isinstance(other, Mapping):
+            return NotImplemented
+        return {
+            k.lower(): v for k, v in self.items()
```

Comment:
The `lower()` call here can be removed, since you already know keys in `self.items()` are in lower case. In fact, you can just compare `self._store` instead.